quickwit-storage/src/debouncer.rs

```diff
@@ -0,0 +1,307 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::hash::Hash;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+
+use async_trait::async_trait;
+use fnv::FnvHashMap;
+use futures::future::{BoxFuture, Shared};
+use futures::{Future, FutureExt};
+use tantivy::directory::OwnedBytes;
+
+use crate::{Storage, StorageResult};
+
+/// The AsyncDebouncer debounces inflight Futures, so that concurrent async request to the same data
+/// source can be deduplicated.
+///
+/// Since we pass the Future potentially to multiple consumer, everything needs to be cloneable. The
+/// data and the future. This is reflected on the generic type bounds for the value V: Clone.
+///
+/// Since most Futures return an Result<V, Error>, this also encompasses the error.
+pub struct AsyncDebouncer<K, V: Clone> {
+    cache: Mutex<FnvHashMap<K, Shared<BoxFuture<'static, V>>>>,
+}
+
+impl<K, V: Clone> Default for AsyncDebouncer<K, V> {
+    fn default() -> Self {
+        Self {
+            cache: Default::default(),
```

Comment:
Nitpick: derive default should work here, no?

Reply:
it's less flexible, because it requires Default on `V`

Reply:
ah yes I keep forgetting it is a bit dumb.